<font size = 5>  Stepwise selection for logistic regression <font> <br>


In [9]:
library(dplyr)

In [2]:
#Load data
df <- read.csv("loan_dataset.csv")
head(df)

Gender Married Dependents Education    Self_Employed ApplicantIncome
1 Male   Yes     1          Graduate     No            4583           
2 Male   Yes     0          Graduate     Yes           3000           
3 Male   Yes     0          Not Graduate No            2583           
4 Male   No      0          Graduate     No            6000           
5 Male   Yes     2          Graduate     Yes           5417           
6 Male   Yes     0          Not Graduate No            2333           
  LoanAmount Loan_Amount_Term Credit_History Property_Area Loan_Status
1 128        360              1              Rural         0          
2  66        360              1              Urban         1          
3 120        360              1              Urban         1          
4 141        360              1              Urban         1          
5 267        360              1              Urban         1          
6  95        360              1              Urban         1

In [3]:
#fit full logistic regression model
full_model <- glm(formula = Loan_Status ~ .,
                  data = df,
                  family = binomial()
                  )
summary(full_model)


Call:
glm(formula = Loan_Status ~ ., family = binomial(), data = df)

Coefficients:
                         Estimate Std. Error z value Pr(>|z|)    
(Intercept)            -2.453e+00  9.280e-01  -2.643  0.00821 ** 
GenderMale              2.639e-01  3.264e-01   0.808  0.41884    
MarriedYes              5.753e-01  2.924e-01   1.968  0.04911 *  
Dependents1            -3.533e-01  3.455e-01  -1.023  0.30644    
Dependents2             2.752e-01  3.777e-01   0.729  0.46618    
Dependents3+            2.100e-01  4.841e-01   0.434  0.66444    
EducationNot Graduate  -3.914e-01  3.018e-01  -1.297  0.19472    
Self_EmployedYes       -1.576e-01  3.515e-01  -0.448  0.65394    
ApplicantIncome         1.485e-05  2.674e-05   0.555  0.57861    
LoanAmount             -3.233e-03  1.710e-03  -1.890  0.05871 .  
Loan_Amount_Term       -8.775e-04  2.024e-03  -0.434  0.66457    
Credit_History          3.648e+00  4.331e-01   8.424  < 2e-16 ***
Property_AreaSemiurban  9.802e-01  3.032e-01   3.233  0.0

<font size = 5> Forward Selection <font>

In [4]:
stepAIC_forward <- MASS::stepAIC(
                                 object = full_model,
                                 direction = "forward"
                                )

Start:  AIC=465.14
Loan_Status ~ Gender + Married + Dependents + Education + Self_Employed + 
    ApplicantIncome + LoanAmount + Loan_Amount_Term + Credit_History + 
    Property_Area



<font size = 5> Backward Selection <font>

In [5]:
stepAIC_backward <- MASS::stepAIC(
                                 object = full_model,
                                 direction = "backward"
                                )


Start:  AIC=465.14
Loan_Status ~ Gender + Married + Dependents + Education + Self_Employed + 
    ApplicantIncome + LoanAmount + Loan_Amount_Term + Credit_History + 
    Property_Area

                   Df Deviance    AIC
- Dependents        3   439.63 461.63
- Loan_Amount_Term  1   437.34 463.34
- Self_Employed     1   437.34 463.34
- ApplicantIncome   1   437.43 463.43
- Gender            1   437.79 463.79
- Education         1   438.79 464.79
<none>                  437.14 465.14
- LoanAmount        1   440.66 466.66
- Married           1   441.04 467.04
- Property_Area     2   450.06 474.06
- Credit_History    1   563.46 589.46

Step:  AIC=461.63
Loan_Status ~ Gender + Married + Education + Self_Employed + 
    ApplicantIncome + LoanAmount + Loan_Amount_Term + Credit_History + 
    Property_Area

                   Df Deviance    AIC
- Loan_Amount_Term  1   439.75 459.75
- Self_Employed     1   439.83 459.83
- ApplicantIncome   1   439.89 459.89
- Gender            1   440.63 460.

<font size =5> Forward Back Selection <font>

In [6]:
stepAIC_both <- MASS::stepAIC(
                              object = full_model,
                              direction = "both"
                             )

Start:  AIC=465.14
Loan_Status ~ Gender + Married + Dependents + Education + Self_Employed + 
    ApplicantIncome + LoanAmount + Loan_Amount_Term + Credit_History + 
    Property_Area

                   Df Deviance    AIC
- Dependents        3   439.63 461.63
- Loan_Amount_Term  1   437.34 463.34
- Self_Employed     1   437.34 463.34
- ApplicantIncome   1   437.43 463.43
- Gender            1   437.79 463.79
- Education         1   438.79 464.79
<none>                  437.14 465.14
- LoanAmount        1   440.66 466.66
- Married           1   441.04 467.04
- Property_Area     2   450.06 474.06
- Credit_History    1   563.46 589.46

Step:  AIC=461.63
Loan_Status ~ Gender + Married + Education + Self_Employed + 
    ApplicantIncome + LoanAmount + Loan_Amount_Term + Credit_History + 
    Property_Area

                   Df Deviance    AIC
- Loan_Amount_Term  1   439.75 459.75
- Self_Employed     1   439.83 459.83
- ApplicantIncome   1   439.89 459.89
- Gender            1   440.63 460.

<font size = 5> Exhaustive Search <font>

In [7]:
#Create dummy variables for categorical columns
df_dummy <- fastDummies::dummy_cols(df,
                                    remove_first_dummy = TRUE,
                                    remove_selected_columns = TRUE
                                    )

In [8]:
bestglm::bestglm(
  Xy = df_dummy %>% select(-Loan_Status),
  family = binomial(),
  IC = "AIC",
  method = "exhaustive"
)

Morgan-Tatar search since family is non-gaussian.



AIC
BICq equivalent for q in (0.813573114090236, 0.911608933189108)
Best Model:
                              Estimate   Std. Error     z value    Pr(>|z|)
(Intercept)                0.667014880 2.955059e-01  2.25719682 0.023995779
LoanAmount                -0.005389829 1.778832e-03 -3.02998141 0.002445688
Dependents_1               1.096111203 3.575025e-01  3.06602397 0.002169259
Dependents_2               0.650077272 3.216224e-01  2.02124354 0.043254561
`Education_Not Graduate`  -0.539143936 3.016116e-01 -1.78754372 0.073849652
Property_Area_Semiurban  -19.742776338 7.553300e+02 -0.02613795 0.979147308

**Full model:** <br>
AIC = 465.14 <br>

The full glm model selects Credit_History, Property_AreaSemiurban and MarriedYes as the most significant predictors. <br>


**Forward Selection model:** <br>
AIC = 465.14 <br>

The forward selection model out of 13 predictors picked 10 of them: <br>
Gender + Married + Dependents + Education + Self_Employed + ApplicantIncome + <br>
LoanAmount + Loan_Amount_Term + Credit_History + Property_Area <br>
and set other coefficients to zero. <br>


**Backward Selection model:** <br>
AIC = 454.72 <br>

The backward selection model picked only 4 predictors and set other coefficients to zero. It is interesting that the backward selection model only picked the predictors that were significant in the full model. This means that the backward selection is more thorough in finding the best features compared to the forward selection model.


**Forward_Backward selection:** <br>
AIC = 454.72 <br>
This model picked the same model as backward selection. This could be because there were not a high number of features so only going backward compared to going backward-forward would not make a difference. <br>


**Exhaustive Serach Selection:** <br>
Obviously this method took longer because it needed to go through all the possible combinations of the features. This model picked LoanAmount, Dependent categories, Education and Property_Area_Semiurban. Choice of Dependent categories and Education were unexpected for me. 